projects/js-packages/components/components/boost-score-graph/annotations-plugin.ts

```diff
@@ -0,0 +1,70 @@
+import uPlot from 'uplot';
+import { Annotation } from '.';
+
+import './style-annotation.scss';
+
+// eslint-disable-next-line jsdoc/require-returns
+/**
+ * Custom tooltips plugin for uPlot.
+ *
+ * @param {Annotation[]} annotations - The periods to display in the tooltip.
+ */
+export function annotationsPlugin( annotations: Annotation[] ) {
+	let containerEl, annotationsContainer;
+
+	/**
+	 * Initialize the plugin
+	 *
+	 * @param {uPlot} u - The uPlot instance.
+	 */
+	function init( u: uPlot ) {
+		containerEl = u.under;
+
+		annotationsContainer = document.createElement( 'div' );
+
+		annotationsContainer.classList.add( 'jb-graph-annotations' );
+
+		const annotationEl = document.createElement( 'div' );
+		annotationEl.classList.add( 'jb-graph-annotations__annotation' );
+
+		annotations.forEach( annotation => {
+			const lineEl = document.createElement( 'div' );
+			lineEl.classList.add( 'jb-graph-annotations__line' );
+			lineEl.addEventListener( 'mouseenter', () => {
+				annotationEl.innerHTML = annotation.text;
+				annotationEl.style.display = 'block';
+				annotationEl.style.left = u.valToPos( annotation.timestamp / 1000, 'x' ) + 'px';
+			} );
+			lineEl.addEventListener( 'mouseleave', () => {
+				annotationEl.style.display = 'none';
+			} );
```

Comment:
My native JS is rusty and I was wondering if it's possible to use one event listener for all.

When looping, add the annotation info as a data property so you can use it on mouse enter for the left position. Then add just one event listener to the parent element.

Just curious, could it be done with one event listener on the parent, instead of for every child? I might be mixing my jQuery days with the way the native JS works 😅

Reply:
I think yes, you might be mixing it with jQuery. Each element will need it's own even listeners.
This is a bit different for the tooltips though. We always show the tooltip for nearest dataset, unless we are not hovering over the graph or highlighting one of these annotations.